### PR TITLE
Include api.consts in api.derive methods

### PR DESCRIPTION
**Pull request by Stefie**
Follow-up of 
https://github.com/polkadot-js/api/pull/1066 
and
https://github.com/polkadot-js/api/pull/1068

### CODE REVIEW

File: packages/api-derive/src/balances/fees.ts [hunk 1]
```diff
@@ -22,25 +20,30 @@ const ZERO = new BN(0);
  *
  * ```javascript
  * api.derive.balances.fees(([creationFee, transferFee]) => {
- *   console.log(`The fee for creating a new account on this chain is ${creationFee} units. The fee required for making a transfer is ${transferFee} units.`);
+ *   // console.log(`The fee for creating a new account on this chain is ${creationFee} units. The fee required for making a transfer is ${transferFee} units.`);
```
Comment (jacogr) [inline packages/api-derive/src/balances/fees.ts]: Why is this commented?

File: packages/api-derive/src/contracts/fees.ts [hunk 2]
```diff
@@ -6,43 +6,114 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
- * Retrieves all the contract fees
+ * @name fees
+ * @returns An object containing the combined results of the queries for
+ * all relevant contract fees as declared in the substrate chain spec.
+ * @example
+ * <BR>
+ *
+ * ```javascript
+ * api.derive.contracts.fees(([creationFee, transferFee]) => {
+ *   // console.log(`The fee for creating a new contract on this chain is ${creationFee} units. The fee required to call this contract is ${transferFee} units.`);
```
Comment (jacogr) [inline packages/api-derive/src/contracts/fees.ts]: As above, the same commented question.

File: packages/api-derive/src/contracts/fees.ts [hunk 3]
```diff
@@ -6,43 +6,114 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
-        rentDepositOffset,
-        transactionBaseFee,
-        transactionByteFee,
-        transferFee,
-        tombstoneDeposit
-      } as DerivedContractFees)),
+    if (api.query.contract && !api.query.contract.rentByteFee) {
+      // Only for 1.0 support. rentByteFee, rentDepositOffset, tombstoneDeposit are not available in substrate 1.0.
+      // @TODO remove this once 1.0 support is dropped
+      const ZERO = new BN(0);
```
Comment (jacogr) [inline packages/api-derive/src/contracts/fees.ts]: Move this const to file-level rather.

File: packages/api-derive/src/balances/fees.ts [hunk 4]
```diff
@@ -22,25 +20,30 @@ const ZERO = new BN(0);
...
-        ? api.queryMulti([
-          api.query.balances.creationFee,
-          api.query.balances.existentialDeposit,
-          api.query.balances.transactionBaseFee,
-          api.query.balances.transactionByteFee,
-          api.query.balances.transferFee
-        ]) as any as Observable<[BN, BN, BN, BN, BN]>
-      : of([ZERO, ZERO, ZERO, ZERO, ZERO])
-    ).pipe(
+    return (api.consts.balances
```
Comment (jacogr) [inline packages/api-derive/src/balances/fees.ts]: (nit-y) comment - this would read better as

```js
return (
  api.consts.balances
    ? of([ ... ])
    : api.queryMulti([...])
)
```

File: packages/api-derive/src/contracts/fees.ts [hunk 5]
```diff
@@ -6,43 +6,114 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
+          queryBase.callBaseFee,
+          queryBase.contractFee,
+          queryBase.createBaseFee,
+          queryBase.creationFee,
+          queryBase.transactionBaseFee,
+          queryBase.transactionByteFee,
+          queryBase.transferFee
+        ])
+      ]) as any as Observable<[[BN, BN, BN], [BN, BN, BN, BN, BN, BN, BN]]>).pipe(
+        map(([[
```
Comment (jacogr) [inline packages/api-derive/src/contracts/fees.ts]: This is a bit difficult to follow, would suggest (if you really want to spread the params into multiple lines), as

```js
map([
  [
    ...
  ],
  [
    ...
  ]
```

File: packages/api-derive/src/contracts/fees.ts [hunk 6]
```diff
@@ -6,43 +6,114 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
+            rentByteFee,
+            rentDepositOffset,
+            tombstoneDeposit
+          ],[
```
Comment (jacogr) [inline packages/api-derive/src/contracts/fees.ts]: If spread over multi lines, don't `], [`, rather `],\n[`

File: packages/api-derive/src/contracts/fees.ts [hunk 7]
```diff
@@ -6,43 +6,114 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
+            tombstoneDeposit,
+            transactionBaseFee,
+            transactionByteFee,
+            transferFee
+          } as DerivedContractFees)),
+        drr()
+      );
+    }
+
+    return (api.consts.contracts
```
Comment (jacogr) [inline packages/api-derive/src/contracts/fees.ts]: Same as the readability ni comment above, i.e.

```js
return (
  api.consts.contracts
    ? of([..])
    : api.queryMulti([...])
)
```

File: packages/api-derive/src/session/info.ts [hunk 8]
```diff
@@ -81,13 +82,16 @@ export function info (api: ApiInterface$Rx) {
...
+        map(createDerived_pre_94),
         drr()
       )
     : (combineLatest([
-      // sessionsPerEra, hardcoded, due to https://github.com/paritytech/substrate/pull/2802/files#diff-5e5e1c3aec9ddfde0a9054d062ab3db9R156
-      of(new BN(6)),
+      // sessionsPerEra:
+      // substrate spec_version >= 94 : get from parameter_types exposed as api.consts
+      // https://github.com/paritytech/substrate/pull/2802/files#diff-5e5e1c3aec9ddfde0a9054d062ab3db9R156
+      // @TODO Switch to `api.constants` after https://github.com/polkadot-js/api/pull/1066 is merged.
```
Comment (jacogr) [inline packages/api-derive/src/session/info.ts]: TODO can be removed now?

File: packages/api-derive/src/balances/fees.ts (hunk 1 above)
Comment (Stefie) [reply packages/api-derive/src/balances/fees.ts]: Oh, a leftover from development. I commented all console.logs to do less scrolling when comparing the results of different nodes. So an oversight.

File: packages/api-derive/src/contracts/fees.ts (hunk 5 above)
Comment (Stefie) [reply packages/api-derive/src/contracts/fees.ts]: I changed it to use the same formatting for the type declarations and the parameters

File: packages/api-derive/src/contracts/fees.ts (hunk 7 above)
Comment (Stefie) [reply packages/api-derive/src/contracts/fees.ts]: changed

File: packages/api-derive/src/contracts/fees.ts (hunk 5 above)
Comment (Stefie) [reply packages/api-derive/src/contracts/fees.ts]: Actually, what about:
```
map(([
  [...],
  [...]
]) => ({
```

File: packages/api-derive/src/balances/all.ts [hunk 9]
```diff
@@ -32,8 +33,8 @@ function calcBalances ([accountId = EMPTY_ACCOUNT, bestNumber = ZERO, [freeBalan
 
   // offset = balance locked at genesis, perBlock is the unlock amount
   const { offset, perBlock } = vesting.unwrapOr(new VestingSchedule());
-  const vestedNow = perBlock.mul(bestNumber);
-  const vestedBalance = vestedNow.gt(offset)
+  const vestedNow: BN = perBlock.mul(bestNumber);
```
Comment (amaurymartiny) [inline packages/api-derive/src/balances/all.ts]: Can be for another PR, but I'm wondering why in api-derive are we putting BNs everywhere? surely BlockNumber and Balance are more semantically correct

File: packages/api-derive/src/balances/fees.ts (hunk 1 above)
Comment (jacogr) [reply packages/api-derive/src/balances/fees.ts]: Phew, was really wondering about that one :)

File: packages/api-derive/src/contracts/fees.ts (hunk 5 above)
Comment (jacogr) [reply packages/api-derive/src/contracts/fees.ts]: I prefer your second version.

Comment (Stefie) [reply packages/api-derive/src/contracts/fees.ts]: That's commited already :)

File: packages/api-derive/src/balances/all.ts (hunk 9 above)
Comment (Stefie) [reply packages/api-derive/src/balances/all.ts]: Good point! But I'd also leave that for another PR since it means touching all files & tests

**Comment by Stefie** [comment]
Anything holding back here? @jacogr @amaurymartiny ?

File: packages/api-derive/src/balances/fees.ts [hunk 10]
```diff
@@ -29,19 +27,25 @@ const ZERO = new BN(0);
 export function fees (api: ApiInterface$Rx) {
   return (): Observable<DerivedFees> => {
     return (
-      // HACK This is a very quick fix to actually get the types in to continue with
-      // other work - so as of now, the fees are completely broken
-      api.query.balances.creationFee
-        ? api.queryMulti([
+      api.consts.balances
```
Comment (amaurymartiny) [inline packages/api-derive/src/balances/fees.ts]: I think this will throw on older nodes `api.consts is undefined`

File: packages/api-derive/src/session/info.ts [hunk 11]
```diff
@@ -13,13 +13,13 @@ import { BlockNumber, Option } from '@polkadot/types';
...
 
-type Result0_94 = [BN, [BN, Option<BlockNumber>, BN, BN, BN]];
+type Result_pre_94 = [BN, [BN, Option<BlockNumber>, BN, BN, BN]];
 type Result = [BN, [BN, BN]];
 
 const ZERO = new BN(0);
 
 // internal helper to just split the logic - take all inputs, do the calculations and combine
-function createDerived0_94 ([bestNumber, [currentIndex, _lastLengthChange, sessionLength, lastEraLengthChange, sessionsPerEra]]: Result0_94): DerivedSessionInfo {
+function createDerived_pre_94 ([bestNumber, [currentIndex, _lastLengthChange, sessionLength, lastEraLengthChange, sessionsPerEra]]: Result_pre_94): DerivedSessionInfo {
```
Comment (amaurymartiny) [inline packages/api-derive/src/session/info.ts]: Let's have the convention everywhere `someThingWhatever_{start_spec}_{end_spec}`, e.g. we already have EventRecord_0_76, here it'd be `createDerived_0_94`

File: packages/api-derive/src/contracts/fees.ts [hunk 12]
```diff
@@ -6,43 +6,107 @@ import { ApiInterface$Rx } from '@polkadot/api/types';
...
+          transactionBaseFee,
+          transactionByteFee,
+          transferFee
+        } as DerivedContractFees)),
+        drr()
+      );
+    }
+
+    return (
+      api.consts.contracts
```
Comment (amaurymartiny) [inline packages/api-derive/src/contracts/fees.ts]: same, possibily undefined

**Comment by amaurymartiny** [comment]
ahah, added some comment 3s late

File: packages/api-derive/src/balances/fees.ts (hunk 10 above)
Comment (amaurymartiny) [reply packages/api-derive/src/balances/fees.ts]: or not, actually

File: packages/api-derive/src/session/info.ts (hunk 11 above)
Comment (jacogr) [reply packages/api-derive/src/session/info.ts]: Sorry :(

Comment (Stefie) [reply packages/api-derive/src/session/info.ts]: Ah, that was the `0`! 💡

File: packages/api-derive/src/balances/fees.ts (hunk 10 above)
Comment (Stefie) [reply packages/api-derive/src/balances/fees.ts]: I tested them with matching nodes for the conditions, so I hope they don't. :)
But you're right.... they should. Even looking at this gives me a "Cannot get property x of undefined"-error.

File: packages/api-derive/src/session/info.ts (hunk 11 above)
Comment (Stefie) [reply packages/api-derive/src/session/info.ts]: https://github.com/polkadot-js/api/pull/1104
